presto-spi/src/main/java/io/prestosql/spi/type/RowType.java

```diff
@@ -371,35 +361,92 @@ else if (!result) {
         return true;
     }
 
-    private static long combineHashCode(long previousFieldHashCode, long currentHash)
+    private static Boolean chainEquals(Boolean previousFieldsEqual, int fieldIndex, MethodHandle nextFieldEqual, Block rightRow, Block leftRow)
+            throws Throwable
     {
-        return 31 * previousFieldHashCode + currentHash;
+        if (previousFieldsEqual == FALSE) {
+            return FALSE;
+        }
+
+        if (leftRow.isNull(fieldIndex) || rightRow.isNull(fieldIndex)) {
+            return null;
+        }
```

Comment:
Does `fieldIndex` correspond to the fields involved in `previousFieldsEqual`? If so, does it mean we're calling equals on the fields without first checking if either of them is null? That would be a change in behavior compared to the current implementation.